File: solutions/python/code-carl/ch02_list/t5_lc0024_swap_pairs.py
"""LeetCode 24. Swap Nodes in Pairs
https://leetcode.cn/problems/swap-nodes-in-pairs/
"""

from typing import Optional

from list_node import ListNode


class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummyHead = ListNode(-1, head)
        curr = dummyHead
        while curr.next is not None and curr.next.next is not None:
            pre, post, tmp = curr.next, curr.next.next, curr.next.next.next
            curr.next = post
            post.next = pre
            pre.next = tmp
            curr = pre

        return dummyHead.next


if __name__ == "__main__":
    from list_utils import build_list, to_list

    solution = Solution()
    assert to_list(solution.swapPairs(build_list([1, 2, 3, 4]))) == [2, 1, 4, 3]
    assert to_list(solution.swapPairs(build_list([1]))) == [1]
    print("All tests passed.")

File: solutions/code-carl/ch02_list/T5_LC0024_SwapPairs.java
/**
 * Practice template for:
 * - LeetCode 24. Swap Nodes in Pairs
 *   https://leetcode.cn/problems/swap-nodes-in-pairs/
 */
public class T5_LC0024_SwapPairs {
    public ListNode swapPairs(ListNode head) {
        ListNode dummyHead = new ListNode(-1, head);
        ListNode curr = dummyHead;
        while (curr.next != null && curr.next.next != null) {
            ListNode pre = curr.next;
            ListNode post = curr.next.next;
            pre.next = post.next;
            post.next = pre;
            curr.next = post;
            curr = pre;
        }
        return dummyHead.next;
    }
}
